File: go/internal/agents/agents_test.go
package agents

import (
	"slices"
	"strings"
	"testing"

	"github.com/osolmaz/slophammer/go/internal/repo"
)

func TestRenderDerivesRunnerAndPackageFacts(t *testing.T) {
	snapshot := testSnapshot(map[string]string{
		"Makefile":                "check:\n\t@true\n",
		"Taskfile.yml":            "check:\n  cmds: [true]\n",
		"justfile":                "check:\n  true\n",
		"go/go.mod":               "module example.com/demo\n",
		"rust/Cargo.toml":         "[workspace]\nmembers = []\n",
		"rust/crate/Cargo.toml":   "[package]\nname = \"demo\"\n",
		"python/pyproject.toml":   "[dependency-groups]\ndev = [\"pytest\"]\n",
		"python/uv.lock":          "version = 1\n",
		"plain/pyproject.toml":    "[project]\nname = \"plain\"\n",
		"npm/package.json":        `{"scripts":{"test":"node --test"}}`,
		"pnpm/package.json":       `{"scripts":{"check":"true"}}`,
		"pnpm/pnpm-lock.yaml":     "lockfileVersion: 9\n",
		"yarn/package.json":       `{"scripts":{"build":"true"}}`,
		"yarn/yarn.lock":          "",
		"bun/package.json":        `{"scripts":{"test":"true"}}`,
		"bun/bun.lock":            "",
		"invalid/package.json":    "{",
		"no-scripts/package.json": `{"name":"empty"}`,
		"nested/Taskfile.yaml":    "check:\n",
		"ignored/Cargo.lock":      "",
		"fixtures/demo/go.mod":    "module example.com/ignored\n",
	})

	evidence := Derive(snapshot)
	if got := strings.Join(evidence.GlobalCommands, ","); got != "make check,task check,just check" {
		t.Fatalf("global commands = %q", got)
	}
	for _, command := range []string{
		"go test ./...",
		"cargo test --workspace",
		"cargo test",
		"uv run pytest",
		"npm test",
		"pnpm run check",
		"yarn run build",
		"bun test",
	} {
		if !slices.Contains(evidence.AllCommands, command) {
			t.Fatalf("all commands do not contain %q: %#v", command, evidence.AllCommands)
		}
	}
	rendered := Render(snapshot)
	if !strings.Contains(rendered, "```sh\nmake check\ntask check\njust check\n```") {
		t.Fatalf("rendered commands missing: %s", rendered)
	}
	if strings.Contains(rendered, "fixtures/demo") {
		t.Fatalf("rendered ignored fixture package: %s", rendered)
	}
	if !strings.Contains(rendered, "- `rust`: `Cargo.toml`") {
		t.Fatalf("rendered package missing: %s", rendered)
	}
}

func TestDeriveUsesWorkspacePackageManagers(t *testing.T) {
	lockSnapshot := testSnapshot(map[string]string{
		"pnpm-lock.yaml":            "lockfileVersion: 9\n",
		"packages/app/package.json": `{"scripts":{"check":"true"}}`,
	})
	if !slices.Contains(Derive(lockSnapshot).AllCommands, "pnpm run check") {
		t.Fatalf("lockfile commands = %#v", Derive(lockSnapshot).AllCommands)
	}

	metadataSnapshot := testSnapshot(map[string]string{
		"package.json":              `{"packageManager":"yarn@4.1.0"}`,
		"packages/app/package.json": `{"scripts":{"build":"true"}}`,
	})
	if !slices.Contains(Derive(metadataSnapshot).AllCommands, "yarn run build") {
		t.Fatalf("metadata commands = %#v", Derive(metadataSnapshot).AllCommands)
	}
}

func TestRenderWithoutEvidenceIsExplicit(t *testing.T) {
	rendered := Render(testSnapshot(nil))
	if !strings.Contains(rendered, "No verification command could be derived") ||
		!strings.Contains(rendered, "No package manifest was detected") {
		t.Fatalf("rendered = %q", rendered)
	}
}

func TestEvaluateAgentInstructionFailures(t *testing.T) {
	tests := []struct {
		name  string
		files map[string]string
		want  []string
	}{
		{name: "missing", files: nil, want: []string{"repo.agents-required"}},
		{name: "empty", files: map[string]string{"AGENTS.md": "# Agents\n"}, want: []string{"repo.agents-empty"}},
		{
			name: "command missing",
			files: map[string]string{
				"AGENTS.md":    "# Agents\n\nKeep changes small and reviewable.\n",
				"package.json": `{"scripts":{"check":"true"}}`,
			},
			want: []string{"repo.agents-commands-required"},
		},
		{
			name: "nested scope",
			files: map[string]string{
				"AGENTS.md":                    "# Agents\n\nRun `npm test` before finishing.\n",
				"packages/app/package.json":    `{"scripts":{"test":"true"}}`,
				"packages/worker/package.json": `{"scripts":{"build":"true"}}`,
			},
			want: []string{"repo.agents-scope-required"},
		},
		{
			name: "invalid and stale block",
			files: map[string]string{
				"AGENTS.md": "# AGENTS.md\n\nKeep changes small and reviewable.\n\n" + StartMarker + "\n## Repository checks\n\n```sh\nnpm test\n```\n" + EndMarker + "\n",
			},
			want: []string{"repo.agents-command-invalid", "repo.agents-stale"},
		},
		{
			name: "unterminated managed block",
			files: map[string]string{
				"AGENTS.md": "# AGENTS.md\n\nKeep changes small and reviewable.\n\n" + StartMarker + "\n",
			},
			want: []string{"repo.agents-stale"},
		},
	}
	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			issues := Evaluate(testSnapshot(test.files))
			got := make([]string, 0, len(issues))
			for _, issue := range issues {
				got = append(got, issue.RuleID)
			}
			if strings.Join(got, ",") != strings.Join(test.want, ",") {
				t.Fatalf("issues = %#v, want %#v", issues, test.want)
			}
		})
	}
}

func TestEvaluateTreatsSupportedCommandAsUseful(t *testing.T) {
	issues := Evaluate(testSnapshot(map[string]string{
		"AGENTS.md": "```sh\ngo test ./...\n```\n",
		"go.mod":    "module example.com/demo\n",
	}))
	if len(issues) != 0 {
		t.Fatalf("issues = %#v", issues)
	}
}

func TestEvaluateAcceptsGeneratedAndNestedInstructions(t *testing.T) {
	snapshot := testSnapshot(map[string]string{
		"packages/app/package.json": `{"scripts":{"check":"true"}}`,
	})
	generated := Render(snapshot)
	snapshot.Files["AGENTS.md"] = repo.File{Path: "AGENTS.md", Content: generated}
	if issues := Evaluate(snapshot); len(issues) != 0 {
		t.Fatalf("generated issues = %#v", issues)
	}

	manual := testSnapshot(map[string]string{
		"AGENTS.md":                 "# Agents\n\nFollow the repository checks before finishing.\n",
		"packages/app/AGENTS.md":    "# App agents\n\nRun `npm run check` before finishing.\n",
		"packages/app/package.json": `{"scripts":{"check":"true"}}`,
	})
	if issues := Evaluate(manual); len(issues) != 1 || issues[0].RuleID != "repo.agents-commands-required" {
		t.Fatalf("manual issues = %#v", issues)
	}
}

func testSnapshot(files map[string]string) repo.Snapshot {
	repoFiles := map[string]repo.File{}
	for filePath, content := range files {
		repoFiles[filePath] = repo.File{Path: filePath, Content: content}
	}
	return repo.NewSnapshot("/repo", repoFiles)
}

File: typescript/tests/agents.test.ts
import { mkdtemp, readFile, symlink, writeFile } from "node:fs/promises";
import os from "node:os";
import path from "node:path";

import { describe, expect, test } from "vitest";

import { agentsFindings, deriveEvidence, renderAgents } from "../src/agents/agents.js";
import { run } from "../src/cli/cli.js";
import { newSnapshot } from "../src/repo/repo.js";

function snapshot(files: Readonly<Record<string, string>>) {
  return newSnapshot(
    "/repo",
    Object.entries(files).map(([filePath, content]) => ({ path: filePath, content }))
  );
}

describe("AGENTS.md evidence", () => {
  test("derives runner and package commands", () => {
    const evidence = deriveEvidence(
      snapshot({
        Makefile: "check:\n\t@true\n",
        "go/go.mod": "module example.com/demo\n",
        "rust/Cargo.toml": "[workspace]\nmembers = []\n",
        "python/pyproject.toml": '[dependency-groups]\ndev = ["pytest"]\n',
        "python/uv.lock": "version = 1\n",
        "web/package.json": '{"scripts":{"check":"true"}}',
        "web/pnpm-lock.yaml": "lockfileVersion: 9\n",
        "fixtures/demo/go.mod": "module example.com/ignored\n"
      })
    );

    expect(evidence.globalCommands).toEqual(["make check"]);
    expect(evidence.allCommands).toEqual(
      expect.arrayContaining([
        "go test ./...",
        "cargo test --workspace",
        "uv run pytest",
        "pnpm run check"
      ])
    );
    expect(evidence.packages.map((area) => area.path)).not.toContain("fixtures/demo");
    expect(renderAgents(snapshot({}))).toContain("No package manifest was detected");
  });

  test("uses workspace package managers", () => {
    const lockEvidence = deriveEvidence(
      snapshot({
        "pnpm-lock.yaml": "lockfileVersion: 9\n",
        "packages/app/package.json": '{"scripts":{"check":"true"}}'
      })
    );
    expect(lockEvidence.allCommands).toContain("pnpm run check");

    const metadataEvidence = deriveEvidence(
      snapshot({
        "package.json": '{"packageManager":"yarn@4.1.0"}',
        "packages/app/package.json": '{"scripts":{"build":"true"}}'
      })
    );
    expect(metadataEvidence.allCommands).toContain("yarn run build");
  });

  test("reports empty, missing commands, scope, invalid commands, and stale evidence", () => {
    expect(agentsFindings(snapshot({})).map((finding) => finding.rule_id)).toEqual([
      "repo.agents-required"
    ]);
    expect(
      agentsFindings(snapshot({ "AGENTS.md": "# Agents\n" })).map((finding) => finding.rule_id)
    ).toEqual(["repo.agents-empty"]);
    expect(
      agentsFindings(
        snapshot({
          "AGENTS.md": "# Agents\n\nKeep changes small and reviewable.\n",
          "package.json": '{"scripts":{"check":"true"}}'
        })
      ).map((finding) => finding.rule_id)
    ).toEqual(["repo.agents-commands-required"]);
    expect(
      agentsFindings(
        snapshot({
          "AGENTS.md": "# Agents\n\nRun `npm test` before finishing.\n",
          "packages/app/package.json": '{"scripts":{"test":"true"}}',
          "packages/worker/package.json": '{"scripts":{"build":"true"}}'
        })
      )
    ).toContainEqual(
      expect.objectContaining({
        rule_id: "repo.agents-scope-required",
        path: "packages/worker/AGENTS.md"
      })
    );

    const stale = [
      "# AGENTS.md",
      "",
      "Keep changes small and reviewable.",
      "",
      "<!-- slophammer:agents:start -->",
      "## Repository checks",
      "",
      "```sh",
      "npm test",
      "```",
      "<!-- slophammer:agents:end -->",
      ""
    ].join("\n");
    expect(
      agentsFindings(snapshot({ "AGENTS.md": stale })).map((finding) => finding.rule_id)
    ).toEqual(["repo.agents-command-invalid", "repo.agents-stale"]);
  });

  test("treats a supported command as useful content", () => {
    expect(
      agentsFindings(
        snapshot({
          "AGENTS.md": "```sh\ngo test ./...\n```\n",
          "go.mod": "module example.com/demo\n"
        })
      )
    ).toEqual([]);
  });

  test("accepts generated evidence", () => {
    const initial = snapshot({ "package.json": '{"scripts":{"check":"true"}}' });
    const generated = renderAgents(initial);
    expect(
      agentsFindings(
        snapshot({
          "AGENTS.md": generated,
          "package.json": '{"scripts":{"check":"true"}}'
        })
      )
    ).toEqual([]);
  });
});

describe("agents CLI", () => {
  test("writes, refuses, forces, previews, and checks", async () => {
    const root = await mkdtemp(path.join(os.tmpdir(), "slophammer-agents-ts-"));
    await writeFile(path.join(root, "Makefile"), "check:\n\t@true\n");

    const preview = await run(["agents", "init", root, "--dry-run"]);
    expect(preview).toMatchObject({ code: 0, stderr: "" });
    expect(preview.stdout).toContain("make check");

    expect(await run(["agents", "init", root])).toMatchObject({ code: 0 });
    const refused = await run(["agents", "init", root]);
    expect(refused.code).toBe(2);
    expect(refused.stderr).toContain("--force");

    await writeFile(path.join(root, "AGENTS.md"), "old\n");
    expect(await run(["agents", "init", root, "--force"])).toMatchObject({ code: 0 });
    expect(await readFile(path.join(root, "AGENTS.md"), "utf8")).toContain("make check");

    const checked = await run(["agents", "check", root, "--format", "json"]);
    expect(checked.code).toBe(0);
    expect(checked.stdout).toContain('"ok": true');
  });

  test("rejects forced symlink replacement", async () => {
    const root = await mkdtemp(path.join(os.tmpdir(), "slophammer-agents-ts-link-"));
    const externalRoot = await mkdtemp(path.join(os.tmpdir(), "slophammer-agents-ts-outside-"));
    const external = path.join(externalRoot, "external.md");
    await writeFile(external, "keep\n");
    await symlink(external, path.join(root, "AGENTS.md"));

    const result = await run(["agents", "init", root, "--force"]);
    expect(result.code).toBe(2);
    expect(result.stderr).toContain("symlink");
    expect(await readFile(external, "utf8")).toBe("keep\n");
  });

  test("rejects invalid agents command arguments", async () => {
    await expect(run(["agents", "check", "--wat"])).resolves.toMatchObject({ code: 2 });
    await expect(run(["agents", "init", ".", ".."])).resolves.toMatchObject({ code: 2 });
    await expect(run(["agents", "wat"])).resolves.toMatchObject({ code: 2 });
  });
});
